src/AMDGPUOffload.cpp

```diff
@@ -0,0 +1,679 @@
+        }
+    }
+
+    uint16_t get_machine() override { return EM_AMDGPU; }
+    uint32_t get_flags() override { return flags; }
+    uint32_t get_version() override { return EV_CURRENT; }
+    void append_dynamic(Section &dynamic) override {
+    }
+
+    uint64_t get_got_entry(Section &got, const Symbol &sym) override {
+        // Check if we already made a got entry for this symbol.
+        for (const Relocation &r : got.relocations()) {
+            if (r.get_symbol() == &sym && r.get_type() == R_AMDGPU_RELATIVE64) {
+                internal_assert(r.get_addend() == 0);
+                return r.get_offset();
+            }
+        }
+
+        uint64_t got_offset = got.contents_size();
+        got.append_contents((uint64_t)0);
+        got.add_relocation(Elf::Relocation(R_AMDGPU_RELATIVE64, got_offset, 0, &sym));
+        return got_offset;
+    }
+
+    bool needs_plt_entry(const Relocation &r) override {
+        return maybe_branch_inst(r.get_type());
+    }
+
+    Symbol add_plt_entry(const Symbol &sym, Section &plt, Section &got, const Symbol &got_sym) override {
+        if (got.contents_empty()) {
+            // The PLT hasn't been started, initialize it now.
+            plt.set_alignment(16);
+
+            std::vector<char> padding(64, (char)0);
+            // TODO: Make a .plt0 entry that supports lazy binding.
+            plt.set_contents(padding.begin(), padding.end());
+        }
+
+        static const uint8_t hexagon_plt1[] = {
+            0x00, 0x40, 0x00, 0x00, // { immext (#0) (Relocation:R_HEX_B32_PCREL_X)
+            0x0e, 0xc0, 0x49, 0x6a, //   r14 = add (pc, ##GOTn@PCREL) }  (Relocation:R_HEX_6_PCREL_X)
+            0x1c, 0xc0, 0x8e, 0x91, //   r28 = memw (r14)
+            0x00, 0xc0, 0x9c, 0x52, //   jumpr r28
+        };
+
+        debug(2) << "Adding PLT entry for symbol " << sym.get_name() << "\n";
+
+        // Add a GOT entry for this symbol.
+        uint64_t got_offset = got.contents_size();
+        got.append_contents((uint32_t)0);
+        got.add_relocation(Elf::Relocation(R_HEX_JMP_SLOT, got_offset, 0, &sym));
+
+        // Add the PLT code.
+        uint32_t plt_offset = plt.get_size();
+        plt.append_contents(hexagon_plt1, hexagon_plt1 + sizeof(hexagon_plt1));
+
+        plt.add_relocation(Relocation(R_HEX_B32_PCREL_X, plt_offset + 0, got_offset, &got_sym));
+        plt.add_relocation(Relocation(R_HEX_6_PCREL_X, plt_offset + 4, got_offset + 4, &got_sym));
+
+        // Make a symbol for the PLT entry.
+        Symbol plt_sym("plt_" + sym.get_name());
+        plt_sym
+            .set_type(Symbol::STT_FUNC)
+            .set_binding(Symbol::STB_LOCAL)
+            .define(&plt, plt_offset, sizeof(hexagon_plt1));
+
+        return plt_sym;
+    }
+
+    Relocation relocate(uint64_t fixup_offset, char *fixup_addr, uint64_t type,
+                        const Elf::Symbol *sym, uint64_t sym_offset, int64_t addend,
+                        Elf::Section &got) override {
```

Comment:
Seems a got_offset needs to be passed into this function from write_shared_object_internal() in elf.cpp by calling get_section_offset(got). That can then be passed into do_relocation for use in performing the R_AMDGPU_GOT* relocations. I am not sure how the Hexagon linker is working.

Also in Elf.cpp `write_shared_object_internal()` the creation of the got section has `got.set_alignment(4);` which should probably be `got.set_alignment(sizeof(addr_t));` as alignment likely needs to be the size of addresses which is 64 bit on AMDGPU.

Reply:
It looks like we simply haven't come across any relocations that need the got offset on Hexagon. There are other relocations that use GOT entries, somehow the Hexagon linker must be able to use the DT_PLTGOT entry in the dynamic library header to hook those up to the GOT at runtime...?

Anyways, it looks like it shouldn't be hard to pass the GOT offset into relocate.  Here's the call:

https://github.com/halide/Halide/blob/master/src/Elf.cpp#L842

You should be able to just add an argument get_section_offset(got) if needed (the offset should be known at this point).